source/_posts/2017-07-22-robot-framework-as-embedded-systems-validation-framework.markdown

```diff
@@ -0,0 +1,350 @@
+---
+author: Piotr Król
+layout: post
+title: "Robot Framework for embedded systems validation - iPXE automation use case"
+date: 2017-07-22 22:42:44 +0200
+comments: true
+categories: apu2 python robotframework validation automation
+---
+
+Recently we started to prepare to [ECC2017](https://ecc2017.coreboot.org/)
+conference. One of topics that we considered was a system for development and
+validation automation.
+
+As maintainers of PC Engines platforms in coreboot we debug and fix quite a lot
+of bugs, but to take full responsibility for our code everything should be
+validated each time we do release. Limited resources lead us to automation and
+as Python enthusiasts we decided to evaluate Robot Framework as first candidate.
+
+When preparing to mentioned conference I found that we lack PXE server from
+which I could run or install needed OSes. Also there was no ready to use
+configuration that gave ability to utilize diskless boot and try recent Linux.
+I started to fight with PXE server configuration, but then I realized that
+without DHCP I have to provide booting information every time typing in iPXE
+shell or integrate custom script. In long term those were not good solutions.
+
+## Robot Framework first try
+
+Project by itself seems to be very popular and at first glance is well
+designed. It gives ability to leverage enormous amount of Python libraries and
+has integrated most important ones. I decided to start with installation in
+virtualenv:
+
+```shell
+[23:00:11] pietrushnic:storage $ virtualenv robot-venv
+Running virtualenv with interpreter /usr/bin/python2
+New python executable in /home/pietrushnic/storage/robot-venv/bin/python2
+Also creating executable in /home/pietrushnic/storage/robot-venv/bin/python
+Installing setuptools, pkg_resources, pip, wheel...done.
+[23:00:29] pietrushnic:storage $ source robot-venv/bin/activate
+(robot-venv) [23:00:36] pietrushnic:storage $ pip install robotframework
+Collecting robotframework
+  Downloading robotframework-3.0.2.tar.gz (440kB)
+    100% |████████████████████████████████| 450kB 2.2MB/s 
+Building wheels for collected packages: robotframework
+  Running setup.py bdist_wheel for robotframework ... done
+  Stored in directory: /home/pietrushnic/.cache/pip/wheels/b4/9b/b2/75d7e5f88f21673eed3472266a1f4e72672328cc174655a1b6
+Successfully built robotframework
+Installing collected packages: robotframework
+Successfully installed robotframework-3.0.2
+```
+
+Verification:
+
+```shell
+(robot-venv) [23:00:46] pietrushnic:storage $ robot --version
+Robot Framework 3.0.2 (Python 2.7.13 on linux2)
+```
+
+## Quick start guide
+
+```
+sudo apt-get install docutils
+git clone https://github.com/robotframework/QuickStartGuide.git
+cd QuickStartGuide
+robot QuickStartGuide.rst
```

Comment:
When I typed:
```
sudo apt-get install docutils
```
I received following response (sorry for polish version of communicates):

```
Czytanie list pakietów... Gotowe
Budowanie drzewa zależności       
Odczyt informacji o stanie... Gotowe
Pakiet docutils jest pakietem wirtualnym zapewnianym przez:
  python3-docutils 0.12+dfsg-1
  python-docutils 0.12+dfsg-1
Należy jednoznacznie wybrać jeden z nich do instalacji.

E: Pakiet docutils nie ma kandydata do instalacji
```
So I decided to type:
```
sudo apt-get install python-docutils 
```
I think adding information about this can be useful, especially when somebody has installed Python 2 and Python 3.